Limit Product/ProductReviews/ProductTerms StoreAPI Per Page

ObliviousHarmony opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

While the API _has_ always supported unbounded queries, in practice, there's not really a good reason to support them. This pull request adds a minimum to those queries and restricts them to `per_page` being between 1 and 100.

Closes #61481
Fixes WOOPLUG-5709

### Breaking Changes:

- Queries to the referenced endpoints are no longer able to be unbounded.
  - Anyone using `per_page=0` will start to receive `400` status codes.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Query `/wc/store/products` endpoint, confirm `per_page` requires a minimum of `1`.
2. Query `/wc/store/products/reviews` endpoint, confirm `per_page` requires a minimum of `1`.
3. Query `/wc/store/products/reviews` endpoint, confirm no `per_page` parameter defaults to `100`.
4. Add the following snippet:
```php
function modify_wc_products_per_page_limit( $endpoints ) {
    // The route pattern for WooCommerce Store API products
    $route = '/wc/store/products';
    
    // Check if the route exists
    if ( ! isset( $endpoints[ $route ] ) ) {
        return $endpoints;
    }
    
    foreach ( $endpoints[ $route ] as $i => $handler ) {
        // Support unbounded queries
        if ( isset( $handler['args']['per_page'] ) ) {
            $endpoints[ $route ][ $i ]['args']['per_page']['minimum'] = 0;
        }
    }
    
    return $endpoints;
}
add_filter( 'rest_endpoints', 'modify_wc_products_per_page_limit' );
```
5. Confirm `OPTIONS` `/wc/store/products` shows a `minimum` of `0` for `per_page`.

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [x] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

Limit `/products` and `/products/reviews` StoreAPI endpoints to 100 results per page.

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

ObliviousHarmony committed:
Limit Products & Reviews StoreAPI Query

ObliviousHarmony committed:
Limit StoreAPI Term Queries

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

Konamiman commented on plugins/woocommerce/src/StoreApi/Routes/V1/Products.php:
```diff
@@ -142,10 +142,10 @@ public function get_collection_params() {
 			'type'              => 'integer',
 			'default'           => 10,
-			'minimum'           => 0,
+			'minimum'           => 1,
```
The pull request description explicitly mentions this as a breaking change. @ObliviousHarmony does this comment change anything (it specifically mentions the Blocks editor)? And same for https://github.com/woocommerce/woocommerce/pull/61755#discussion_r2488403788

ObliviousHarmony replied:
I'm a big fan of this bot, I'm going to need to re-think the compatibility issues with this change.

nerrad commented on plugins/woocommerce/src/StoreApi/Routes/V1/AbstractTermsRoute.php:
```diff
@@ -37,7 +37,8 @@ public function get_collection_params() {
 		$params['per_page'] = array(
 			'description'       => __( 'Maximum number of items to be returned in result set. Defaults to no limit if left blank.', 'woocommerce' ),
```
Should this text be updated?
```suggestion
			'description'       => __( 'Maximum number of items to be returned in result set.', 'woocommerce' ),
```

ObliviousHarmony replied:
Yep!

I _am_ considering removing the `maximum` here and allowing for unbounded queries. It doesn't have a `default` configured so the expectation is currently that, unless explicitly desired, it will return all terms. There's no great way to give errors to notify the consumer that the query is no longer valid.

The attribute terms (options you can select for a given attribute) likely won't ever be so much that it breaks anything. Some folks have a lot of product categories, however, we're only talking about a tiny string and I doubt it'll be easy to exhaust memory with categories. I can see a lot of good use-cases for it too.

ObliviousHarmony replied:
Since I've decided to allow for these queries to remain unbounded, there's no longer a need to change the description.

ObliviousHarmony committed:
Allow Unbounded Term Requests

Since the default is `0` and most requests would have been
written with that expectation, the silent backward
compatibility break isn't worth the improvement.

In practice, term endpoints aren't going to consume much memory.
There is a strong case for always fetching all of the terms, such as
to display options in a user interface or cache them somewhere.

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

ObliviousHarmony committed:
Documentation Fix

ObliviousHarmony committed:
Fixed Changelog

Aggressive bot!

ObliviousHarmony committed:
Mardown Style Fix

ObliviousHarmony commented:
Looks like WP Ocean wouldn't be affected by this. It provides a special `posts-per-page=all` query parameter that causes the underlying loop query to load all of the products.

kraftbj committed:
Merge branch 'trunk' into add/61481-minimum-query-per-page

ObliviousHarmony commented on plugins/woocommerce/src/StoreApi/Routes/V1/ProductReviews.php:
```diff
@@ -184,10 +184,10 @@ public function get_collection_params() {
 			'type'              => 'integer',
 			'default'           => 10,
-			'minimum'           => 0,
+			'minimum'           => 1,
```
All of our queries have been migrated to use pagination or have a different UX.

ObliviousHarmony commented on plugins/woocommerce/src/StoreApi/Routes/V1/AbstractTermsRoute.php:
```diff
@@ -37,7 +37,8 @@ public function get_collection_params() {
 			'type'              => 'integer',
-			'minimum'           => 0,
+			'default'           => 100,
+			'minimum'           => 1,
```
I reverted this change because term endpoints aren't a heavy request path and it's more likely people might want all of given kind of term than it is they'd want all of the products on a store.
